sw/host/virtual_uart.c

```diff
@@ -0,0 +1,215 @@
+    u_poll_period = thread_arg->u_poll_period;
+
+    /* Open the /dev/mem file */
+    fd = open("/dev/mem", O_RDWR | O_SYNC);
+    if (fd == -1) {
+	    printf("ERROR: Cannot open device file\n");
+	    goto end;
+    }
+
+    /* Compute the page aligned offset  */
+    pa_offset = paddr+offset & ~(sysconf(_SC_PAGE_SIZE) - 1);
+    
+    /* Get the virtual address */
+    map = mmap(NULL, length + paddr+offset - pa_offset, PROT_READ | PROT_WRITE, MAP_SHARED, fd, pa_offset);
+    if (map == MAP_FAILED) {
+        printf("ERROR: Map failed\n");
+        goto end;
+    }
+
+    while (1) {
+        
+        /* Poll on the status flag TX full */
+        status_tx_full = ((*(uint8_t *)(map+8) & TX_FULL_BIT_MASK) >> 3);
+        while( !status_tx_full ) {
+            usleep(u_poll_period);
+            status_tx_full = ((*(uint8_t *)(map+8) & TX_FULL_BIT_MASK) >> 3);
+        }
+
+        /* Directly read the data in the TX register, no need to check the status */
+        c = *(char *) (map+4);
+        putchar(c);
+
+        /* ACK the interrupt */
+        *(uint32_t *) (map+16) = 0x000000FF;
+    }
+
+    end:
+        if(map) 
+            munmap(map, length + offset - pa_offset);
+        if (fd)
+            close(fd);
+        return NULL;
+}
+
+
+void disable_buffering () 
+{
+    struct termios t;
+    tcgetattr(STDIN_FILENO, &t);           // Get current terminal settings
+    t.c_lflag &= ~ICANON;                  // Disable canonical mode (line buffering)
+    t.c_lflag |= ECHO;                     // Enable echo so typed characters appear
+    tcsetattr(STDIN_FILENO, TCSANOW, &t);  // Apply new terminal settings
+}
+
+void enable_buffering () 
+{
+    struct termios t;
+    tcgetattr(STDIN_FILENO, &t);           // Get current terminal settings
+    t.c_lflag |= ICANON;                   // Enable canonical mode
+    t.c_lflag |= ECHO;                     // Ensure echo is re-enabled
+    tcsetattr(STDIN_FILENO, TCSANOW, &t);  // Apply new terminal settings
+}
+
+
+void help (char * ex_name)
+{
+    printf("------------------------------ VIRTUAL UART ----------------------------------- \n"); 
+    printf("Usage: %s <uart_paddr> <uart_length> <offset> <u_poll_period>\n", ex_name); 
+    printf("    uart_paddr    : UART physical address in hex 0x... (PCIe BAR)\n");
+    printf("    uart_length   : UART register length in byte, always 20\n");
+    printf("    offset        : UART offset, always 0\n");
+    printf("    u_poll_period : Poll period in microseconds, 10 is good\n");
+    printf("------------------------------------------------------------------------------- \n"); 
+}
+
+
+int main ( int argc, char *argv[] )
+{
+
+    pthread_t read_thread;
+    pthread_t write_thread;
+
+    write_thread_arg_t * write_thread_arg = (write_thread_arg_t *) malloc (sizeof(write_thread_arg_t));
+    read_thread_arg_t * read_thread_arg = (read_thread_arg_t *) malloc (sizeof(read_thread_arg_t));
+
+    if ( argc < 5 ) {
```

Comment:
5 mandatory args are a bit too much:
- We should insert a default value for most, or possibly, all args; 
- The address is always the same after all
- Conversely and if necessary, for some config-specific parameters, e.g. AXI_DATA_WIDTH etc, the virtual UART hardware itself could expose some CSR to software

Reply:
Removed offset and add default values for uart_length and u_poll_period (20, 10)